improve GUI consistency & UX for data-defined style

vmora opened the pull request:
fixes issue 9881 

This PR replaces #1833

vmora committed:
improve GUI consistency & UX for data-defined style
fixes #9881

nyalldawson commented:
@vmora - thanks, will review ASAP

nyalldawson commented on src/gui/symbology-ng/qgssymbollayerv2widget.h:
```diff
@@ -17,10 +17,15 @@
+
 #include <QWidget>
+#include <QMap>
+#include <QMap>
```
duplicate line

nyalldawson commented on src/gui/symbology-ng/qgssymbollayerv2widget.h:
```diff
@@ -17,10 +17,15 @@
 
 class QgsSymbolLayerV2;
 class QgsVectorLayer;
+class QgsDataDefinedButton;
```
shouldn't be required given that the whole header is included

nyalldawson commented on src/gui/symbology-ng/qgssymbollayerv2widget.cpp:
```diff
@@ -45,6 +45,28 @@
+      symbolLayer()->setDataDefinedProperty( it.key(), it.value()->currentDefinition() );
+  }
+  emit changed();
+}
```
@vmora Here's a simpler approach: 
```C++
void QgsSymbolLayerV2Widget::registerDataDefinedButton( QgsDataDefinedButton * button, const QString & propertyName, QgsDataDefinedButton::DataType type, const QString & description )
{
  QgsDataDefined dd( symbolLayer()->dataDefinedProperty( propertyName ) );
  button->init( mVectorLayer, &dd, type, Qt::escape( description ) );
  button->setProperty( "propertyName", propertyName );
  connect( button, SIGNAL( dataDefinedChanged( const QString& ) ), this, SLOT( updateDataDefinedProperty() ) );
  connect( button, SIGNAL( dataDefinedActivated( bool ) ), this, SLOT( updateDataDefinedProperty() ) );
}

void QgsSymbolLayerV2Widget::updateDataDefinedProperty()
{
  QgsDataDefinedButton* button = qobject_cast<QgsDataDefinedButton*>( sender() );
  QString propertyName = button->property( "propertyName" ).toString();
  if ( button->isActive() )
  {
    symbolLayer()->setDataDefinedProperty( propertyName, button->currentDefinition() );
  }
  else
  {
    symbolLayer()->removeDataDefinedProperty( propertyName );
  }
  emit changed();
}
```

This avoids the need to remove all properties and rebuild them, and means that mDataDefinedPropertyButtons is no longer required.

nyalldawson commented:
@vmora Looks much better, thanks. See my note about removing mDataDefinedPropertyButtons - otherwise this is good to merge. A better (more descriptive) commit log message would also be appreciated. Great work!

vmora committed:
improve GUI consistency & UX for data-defined style
data defined buttons added to the gui
improves the design from the previous commit

vmora commented:
@nyalldawson thanks for the suggestion, it really improves the design.

I triple checked my registered data-defined buttons and noticed that the marker outline style for ellipse was not registered. This is because the expression was not set in the code before, so I fixed that.

NathanW2 commented:
How close to merge is this?

On Wed, 21 Jan 2015 9:23 pm Vincent Mora <notifications@github.com> wrote:

> @nyalldawson <https://github.com/nyalldawson> thanks for the suggestion,
> it really improves the design.
>
> I triple checked my registered data-defined buttons and noticed that the
> marker outline style for ellipse was not registered. This is because the
> expression was not set in the code before, so I fixed that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/1842#issuecomment-70821551>.
>

vmora commented:
qgis_ziplayertest fails on Travis, but seems unrelated.

nyalldawson commented:
Manually squashed and merged - thank you!
